drivers: sensor: zephyr_thermistor: refactor driver

gmarull opened the pull request:
Refactor driver to align a bit more with its Linux counterpart, ie,
ntc_thermistor. This driver did quite a few _unconventional_ things,
like using "zephyr," compatibles, a dedicated node for pre-computed
compensation table (referenced by the actual pseudo-device node), etc.
The comparison helper function should likely be simplified as well (to
avoid the need for custom wrapper for bsearch), but this can be done
later.

In this refactor, each thermistor gets a compatible, e.g. "epcos,xxxx".
Compatibles are known by the driver, so are compensation tables. This
simplifies devicetree files. There's no need to bother about
compensation tables in **every** board file if Zephyr supports a certain
NTC model.

In general, we should respect Linux bindings, which in the end influence
how drivers are implemented. In this case, this principle resulted in
simplified, easier to use code.

For future developers, this is how support for a new NTC can be added:

1. Add a new compensation table, named `comp_$DT_DRV_COMPAT$`
2. Add to the end of the driver:

```c
 #undef DT_DRV_COMPAT
 #define DT_DRV_COMPAT vnd_model
 #define DT_INST_FOREACH_STATUS_OKAY_VARGS(NTC_THERMISTOR_DEV_INIT,
                                           DT_DRV_COMPAT)
```
3. In driver's Kconfig make sure it depends on
   DT_HAS_$DT_DRV_COMPAT$_ENABLED

Note: $X$ means _value_ of X.

Also added support for a generic NTC, `ntc-thermistor-generic`. In this case,
the compensation table is provided via devicetree. Note that DT property
is prefixed with `zephyr,`, because while hardware related, it is linked
to a particular software implementation.

pdgendt commented:
There are other active PRs related to this:
- https://github.com/zephyrproject-rtos/zephyr/pull/56066
- https://github.com/zephyrproject-rtos/zephyr/pull/57341

Maybe this should be discussed during a dev meeting?

CC @or-rhssk @asemjonovs

pdgendt commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -0,0 +1,155 @@
+	const struct ntc_config ntc_cfg;
+};
+
+static const struct ntc_compensation comp_epcos_b57861s0103a039[] = {
```
Do we want to have tables for every possible NTC thermistor?

Also, doing it like this dictates the range and steps.

gmarull replied:
I think it's a good idea to simplify usage. Imagine we have 10 boards with the same NTC in-tree. We'll basically need to maintain 10 copies of the compensation table. I also tend to trust Linux drivers more than Zephyr... :laughing:

gmarull commented:
> There are other active PRs related to this:
> 
> * [drivers: sensor: Add generic ADC thermistor sensor driver #56066](https://github.com/zephyrproject-rtos/zephyr/pull/56066)
> * [sensor: thermistor: improve temperature calculation #57341](https://github.com/zephyrproject-rtos/zephyr/pull/57341)
> 
> Maybe this should be discussed during a dev meeting?
> 
> CC @or-rhssk @asemjonovs

Thanks for the references, @pdgendt. #56066 looks pretty clean and simpler, even though it should replace existing driver as it's clearly overlapping. I think what would be nice is to have a universal driver that can either take a known thermistor compatible (so that you just specify 'vnd,model'), or a generic one, like `ntc-thermistor`. For the first case, driver would take a known compensation table within driver file, for the second, take the table from DT. This would allow us to keep Linux compatibility, simplify usage for supported NTCs, and allow using any random NTC if not yet supported.

https://github.com/zephyrproject-rtos/zephyr/pull/57341 look like a bug fix/improvement, so not strictly related. In this PR I'm just refactoring DT usage, but driver behavior should stay the same.

gmarull committed:
drivers: sensor: zephyr_thermistor: refactor driver

Refactor driver to align a bit more with its Linux counterpart, ie,
ntc_thermistor. This driver did quite a few _unconventional_ things,
like using "zephyr," compatibles, a dedicated node for pre-computed
compensation table (referenced by the actual pseudo-device node), etc.
The comparison helper function should likely be simplified as well (to
avoid the need for custom wrapper for bsearch), but this can be done
later.

In this refactor, each thermistor gets a compatible, e.g. "epcos,xxxx".
Compatibles are known by the driver, so are compensation tables. This
simplifies devicetree files. There's no need to bother about
compensation tables in **every** board file if Zephyr supports a certain
NTC model.

In general we should respect Linux bindings, which in the end influence
how drivers are implemented. In this case, this principle resulted in
simplified, easier to use code.

For future developers, this is how support for a new NTC can be added:

1. Add to the end of the driver:

```c
 #undef DT_DRV_COMPAT
 #define DT_DRV_COMPAT vnd_model

 static __unused const struct ntc_compensation comp_vnd_model[] = {
     { x, y },
     ...,
 };

 #define DT_INST_FOREACH_STATUS_OKAY_VARGS(NTC_THERMISTOR_DEV_INIT,
                                           DT_DRV_COMPAT, comp_vnd_model)
```
3. In driver's Kconfig make sure it depends on
   DT_HAS_$DT_DRV_COMPAT$_ENABLED

Note: $X$ means _value_ of X.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: sensor: zephyr_thermistor: align connection type with Linux

It looks like the Zephyr thermistor driver bindings were half-copied
from Linux ntc-thermistor. Zephyr principle is to maintain compatibility
with Linux, when possible, so there's no reason to deviate here. Convert
the connection type from a custom enum to a boolean, as Linux does.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

asemjonovs commented:
@gmarull - These changes look very similar to my initial draft of adding the thermistor driver to Zephyr.  It very much mimicked Linux's implementation initially.  After much feedback from reviewers it evolved into what it is right now.  Some of the discussions on how the thermistor table moved from code to device tree can be found in the [PR](https://github.com/zephyrproject-rtos/zephyr/pull/54947) and discord [thread](https://discord.com/channels/720317445772017664/1090017807527317654/1090017810056495226)

gmarull commented:
> @gmarull - These changes look very similar to my initial draft of adding the thermistor driver to Zephyr. It very much mimicked Linux's implementation initially. After much feedback from reviewers it evolved into what it is right now. Some of the discussions on how the thermistor table moved from code to device tree can be found in the [PR](https://github.com/zephyrproject-rtos/zephyr/pull/54947) and discord [thread](https://discord.com/channels/720317445772017664/1090017807527317654/1090017810056495226)

I think we need to revisit some of the decisions that were made. First, there's no reason to not align with Linux bindings here (see [this](https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#always-check-for-existing-bindings)). I don't really get why we have, for example, coefficients in a separate node, referenced via a phandle. We can have a multi-compatible driver that can support any number of known thermistors, without the need to keep compensation tables in DT (as in Linux). We could also introduce a `ntc-thermistor-generic` compatible that just extends `ntc-thermistor` with a lookup table property, to allow usage of any random NTC. I've added the topic to the next Arch WG so we can discuss.

gmarull committed:
drivers: sensor: ntc_thermistor: add support for generic NTC

Add support for a generic NTC, `ntc-thermistor-generic`. In this case,
the compensation table is provided via devicetree. Note that DT property
is prefixed with `zephyr,`, because while hardware related, it is linked
to a particular software implementation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

pdgendt commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -0,0 +1,171 @@
+	const struct ntc_config ntc_cfg;
+};
+
+static const struct ntc_compensation comp_epcos_b57861s0103a039[] = {
+	{ -25, 146676 },
+	{ -15, 78875 },
+	{ -5, 44424 },
+	{ 5, 26075 },
+	{ 15, 15881 },
+	{ 25, 10000 },
+	{ 35, 6488 },
+	{ 45, 4326 },
+	{ 55, 2956 },
+	{ 65, 2066 },
+	{ 75, 1474 },
+	{ 85, 1072 },
+	{ 95, 793 },
+	{ 105, 596 },
+	{ 115, 454 },
+	{ 125, 351 },
+};
```
Maybe move this to where you specify the compatible below?
```c
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT epcos_b57861s0103a039
DT_INST_FOREACH_STATUS_OKAY_VARGS(NTC_THERMISTOR_DEV_INIT, DT_DRV_COMPAT)
```

gmarull replied:
done thanks

or-rhssk commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -0,0 +1,168 @@
+				.connected_positive = DT_INST_PROP(inst, connected_positive),      \
+				.type = {                                                          \
+					.comp = (const struct ntc_compensation *)comp_table,       \
+					.n_comp = ARRAY_SIZE(comp_table),                          \
```
@gmarull stumbled on this line when trying to rebase #57341 on top of this.

I think the array size is a bit misleading, since it refers to the `comp_table` array which has 2x members (two `uint32_t`s for each `struct ntc_compensation`) and not to the `comp` member. I assume that's also the reason for changes in the `bsearch` helper function (`element_idx == type->n_comp - 1` -> `element_idx == (type->n_comp / 2) - 1`). But there's still places where `n_comp` is not updated to account for this change (e.g. in `ntc_lookup_comp`).

Wouldn't it be cleaner to simply set `n_comp` to the real `comp_table` size, like so?
```c
.n_comp = ARRAY_SIZE(comp_table) / 2,
```
